### DISCUSSION TITLE
Query execution time increases when the number of concurrent users increases

### DISCUSSION THREAD

**Discussion by vishalratnam-cliq-scm** [vesoft-inc/nebula#4223] [Q&A]
We're trying to load test (using [k6](https://k6.io/) testing tool) nebula graph DB using the [nebula-go](https://github.com/vesoft-inc/nebula-go) client. 
We have created 1 global connection pool when the application starts with 1 host address pointing to our cluster (nebula cluster contains 3 graphd services, 3 metad services and 3 storaged services). 
We create a session for each incoming request on our API, that executes a query and returns the data. After the API is done, we close the session. 
The response time when we were trying to do the same with 1 virtual user on k6, the query execution time is 20ms. When we hit the same API with 10 virtual users with 10 concurrent requests from each user, the query execution time increases to ~100ms. 

Is there any reason for this increase in execution time? Do we need to create a new pool for each request? Can't we use the same pool of connections when the load increases?

**Comment by xiajingchun** [comment]
Session is also expensive to create. It's highly recommened that you keep one session for each virtual user. Give it a try.
Here's a k6 plugin we use to test Nebula, which might interest you as well: https://github.com/vesoft-inc/k6-plugin

**Reply by vishalratnam-cliq-scm** [reply]
Hey @xiajingchun, understood. Our use case is API-based. We have a GraphQL endpoint that queries nebula every time it's hit. 
Do you suggest using the same session for each API call? We tried this and got `broken pipe` error when we increased the number of users. Hence we went with creating a new session for each API call.
How do we create sessions efficiently such that every time our API is hit, the nebula queries are fast?

**Reply by xiajingchun** [reply]
Session is not thread-safe, therefore, you can't reuse one single session like this. Similar to a connection pool, you may initialize and maintain a pool of session, then, each API call simply gets the session from the pool, and return it after the query is done, without truly releasing a session.

**Reply by vishalratnam-cliq-scm** [reply]
@xiajingchun is there any function/API provided by nebula/ nebula-go to create the session pool? 
I'm aware that nebula-go provides a method for creating connection pool, but haven't come across session pools.

**Reply by xiajingchun** [reply]
I'm afraid no, but it won't be too complicated to create one by yourself :)

**Reply by jamieliu1023** [reply]
@chvr203 Welcome to contribute back to the community if you finish it ;)

**Reply by akhilravuri-tul-scm** [reply]
@xiajingchun so that means without session pool we cannot make concurrent reads from graphDB?

**Reply by xiajingchun** [reply]
That's not what I meant :). In chvr203's example, concurrent reads is already possible but at a cost of opening/releasing sessions for every single request, and that hurts the performance. Session pool is a good practice, not a must. You can have you own way to manage sessions, just keep in mind it's not thread safe.

**Reply by akhilravuri-tul-scm** [reply]
@xiajingchun but when we try to create 300 sessions concurrently it returns 
```Internal error occured. Please contact the system administrator. failed to get connection: No valid connection in the idle queue and connection number has reached the pool capacity```

We have modified our pool capacity(max connections) to 1million.

**Reply by xiajingchun** [reply]
@akhilravuri2 that's strange, would you mind to put here some code snippets so we can take a look?

**Reply by akhilravuri-tul-scm** [reply]
```
	hostAddr := nebula_sdk.HostAddress{Host: address, Port: port}
	hostList := []nebula_sdk.HostAddress{hostAddr}
	// Create configs for connection pool using default values
	poolConfig := nebula_sdk.GetDefaultConf()
	// Initialize connection pool
	connPool, err := nebula_sdk.NewConnectionPool(hostList, poolConfig, nebula_sdk.DefaultLogger{})
```
This connection pool is stored globally.

When a call is made from postman to the below end point we create a session and execute a query and process the result and send the result back.

```
func GraphQLQuery(){
session, err := connPool.GetSession(username, password)
defer session.Release()

totalQuery := "Fetch Query"

res, err = session.Execute(totalQuery)
if err != nil {
	return nil, err
}

if res.IsSucceed() {
	return res, nil
}
}
```
This was working fine with postman but when we started load testing it with k6... where ```vus: 11,``` It started giving No valid connection error.

Then we changed the poolConfig.MaxConnPoolSize = 1000000 and tried to execute it with Vus:300 and failed with No valid connection.

But when we tried with 200 it was working fine.

@xiajingchun Please let me know if need anything.

**Reply by akhilravuri-tul-scm** [reply]
```
import http from 'k6/http';

export const options = {
    vus: 300,
    duration: '30s',
    };

export default function() {
const query = `
query {
    GraphQLQuery (input:{ID:"123-abcd"}){
        name
        place
    }
}`;

const headers = {
  'Content-Type': 'application/json',
};

const res = http.post('http://localhost:8000/query', JSON.stringify({ query: query }), {
  headers: headers,
});
}
```

This is the k6 test script.

**Reply by HarrisChu** [reply]
how do you change `connPool.MaxConnPoolSize = 1000000 `
when `NewConnectionPool`, it would initialize the connection pool.
so just config the `MaxConnPoolSize` in poolConfig

```
poolConfig := nebula_sdk.GetDefaultConf()
poolConfig.MaxConnPoolSize = 1000000
connPool, err := nebula_sdk.NewConnectionPool(hostList, poolConfig, nebula_sdk.DefaultLogger{})
```

**Reply by HarrisChu** [reply]
could you verify the connection via `netstat |grep 9669|wc -l`

**Reply by akhilravuri-tul-scm** [reply]
> how do you change `connPool.MaxConnPoolSize = 1000000 ` when `NewConnectionPool`, it would initialize the connection pool. so just config the `MaxConnPoolSize` in poolConfig
> 
> ```
> poolConfig := nebula_sdk.GetDefaultConf()
> poolConfig.MaxConnPoolSize = 1000000
> connPool, err := nebula_sdk.NewConnectionPool(hostList, poolConfig, nebula_sdk.DefaultLogger{})
> ```

Yes..we use poolConfig.MaxConnPoolSize = 1000000 only.

**Reply by vishalratnam-cliq-scm** [reply]
Hi @HarrisChu,
The output of `netstat |grep 9669|wc -l` changes as we increase the number of virtual users from K6. 
We set the `poolConfig.MaxConnPoolSize` to `100000`. After increasing the number of virtual users from 10 to 250, the output of `netstat |grep 9669|wc -l` increased, but was never equal to the number of virtual users. Please check the below stats:

Single Pod:
```
1 VUs - 1 Connection
10 VUs - 8 Connections
50 VUs - 40 Connections
100 VUs - 81 Connections and 94, 97
150 VUs - 97 Connections
200 VUs - 166 Connections
250 VUs - 189 Connections, Errors at this point: Error: dial tcp 15.206.141.229:80: socket: too many open files
```
2 Pods:
```
1 VUs -  1 on 1 pod
10 VUs - 4 on each pod
50 VUs - 20 on each pod
100 VUs - 40 on pod 1, 41 on pod2
150 VUs - 50 on pod 1, 52 on pod2
200 VUs - 50 on pod 1, 52 on pod2
250 VUs - 65 on each pod, Errors at this point: Error: dial tcp 15.206.141.229:80: socket: too many open files
```

**Reply by xiajingchun** [reply]
could you check the result of 
`ulimit -n`
more details at: https://docs.nebula-graph.io/3.1.0/5.configurations-and-logs/1.configurations/6.kernel-config/#ulimit_-n

**Reply by akhilravuri-tul-scm** [reply]
@xiajingchun 
`
sh-4.4# ulimit -n
`
`
1048576
`

**Reply by akhilravuri-tul-scm** [reply]
@xiajingchun why current nebula-go client connection pool does not have MaxWaitTime(time to wait till a session is available) kind of option.

Because if we use 1k we are getting `too many open files` and if we use 100(MaxConPool) and 1k(Vus) we are getting `No valid connection in the idle queue and connection number has reached the pool capacity`

**Comment by HarrisChu** [comment]
I am not sure, but suppose topology as below:
![image](https://user-images.githubusercontent.com/1726587/167979895-e92e024f-31ca-4336-b472-8d4fd23bc5f5.png)

1. why 150 VUs using 97 Connections
maybe the http server is slow. i.e. when http server send 98th query, the 1st query is finished and then return the connection to pool. So no need to create a new connection.

2. `too many open files`
when raise the error? Immediately or after several seconds?
I'm not sure if the http server is in the same host with nebula. maybe you can verify 
```
netstat |grep {your http server port}|wc -l
# or
sudo lsof -p {graph pid}|wc -l
sudo lsof -p {storage pid}|wc -l
sudo lsof -p {http server pid}|wc -l
```

**Reply by akhilravuri-tul-scm** [reply]
@HarrisChu error is raised immediately if concurrent connection count is more than 65.

http server and nebula are on different host(node in cluster)
